packages/@react-spectrum/color/test/HexColorField.test.js

```diff
@@ -0,0 +1,223 @@
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('');
+
+    hexColorField.focus();
+    hexColorField.blur();
+    expect(hexColorField.value).toBe('');
+  });
+
+  it.each`
+    Name                      | props
+    ${'3-length hex string'}  | ${{defaultValue: '#abc'}}
+    ${'6-length hex string'}  | ${{defaultValue: '#aabbcc'}}
+    ${'Color object'}         | ${{defaultValue: new Color('#abc')}}
+  `('should accept $Name as value', function ({props}) {
+    renderComponent(props);
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#AABBCC');
+  });
+
+  it.each`
+    Name                   | props
+    ${'custom min value'}  | ${{defaultValue: '#aaa', minValue: '#bbb'}}
+    ${'custom max value'}  | ${{defaultValue: '#ccc', maxValue: '#bbb'}}
+  `('should clamp initial value provided to $Name', function ({props}) {
+    renderComponent(props);
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#BBBBBB');
+  });
+
+  it('should handle uncontrolled state', function () {
+    const onChangeSpy = jest.fn();
+    renderComponent({defaultValue: '#abc', onChange: onChangeSpy});
+
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
+    fireEvent.change(hexColorField, {target: {value: 'cba'}});
+    expect(hexColorField.value).toBe('cba');
+    expect(onChangeSpy).toHaveBeenCalledWith(new Color('#cba'));
+
+    hexColorField.blur();
+    expect(hexColorField.value).toBe('#CCBBAA');
+  });
+
+  it('should handle controlled state', function () {
+    const onChangeSpy = jest.fn();
+    renderComponent({value: '#abc', onChange: onChangeSpy});
+
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
+    fireEvent.change(hexColorField, {target: {value: 'cba'}});
+    expect(hexColorField.value).toBe('cba');
+    expect(onChangeSpy).toHaveBeenCalledWith(new Color('#cba'));
+
+    hexColorField.blur();
+    expect(hexColorField.value).toBe('#AABBCC');
+  });
+
+  it('should revert back to last valid value', function () {
+    const onChangeSpy = jest.fn();
+    renderComponent({defaultValue: '#abc', onChange: onChangeSpy});
+    const hexColorField = screen.getByLabelText('Primary Color');
+    expect(hexColorField.value).toBe('#AABBCC');
+
+    hexColorField.focus();
+    fireEvent.change(hexColorField, {target: {value: 'xyz'}});
+    expect(hexColorField.value).toBe('xyz');
+    expect(onChangeSpy).not.toHaveBeenCalled();
+
+    hexColorField.blur();
+    expect(hexColorField.value).toBe('#AABBCC');
+  });
+
+  it.each`
+    Name                                | expected      | action
+    ${'increment with arrow up key'}    | ${'#AAAAAE'}  | ${(el) => fireEvent.keyDown(el, {key: 'ArrowUp'})}
```

Comment:
Will do! I wrote these tests following `useSpinButton.test.js` that only fired `keyDown` events

Reply:
yeah, this is something we're starting to be better about, when we first started this code base we just did the events that we needed, however, we ran into some weird cases where tests were failing or passing when they shouldn't because the events didn't match how a real user would interact